### ISSUE TITLE
harmonic.salience - masked arrays too slow?

### ISSUE THREAD

**Issue by rabitt** [librosa/librosa#456]
I'm finding that when running the harmonic.salience function, the [aggregate](https://github.com/librosa/librosa/blob/master/librosa/core/harmonic.py#L107) function is running slower than it should (longer than the interpolator in some cases). I found [this issue](http://stackoverflow.com/questions/5760668/python-numpy-masked-arrays-are-very-slow) reported before. Have you run into this before? I remember you asked me to switch to masked arrays precisely for efficiency.

**Comment by bmcfee** [comment]
> Have you run into this before? I remember you asked me to switch to masked arrays precisely for efficiency.

I hadn't, no.  My other reason for suggesting masked arrays is that they make it trivial to support arbitrary aggregators on a selected index.  The other way I can think of to do this is to fill all non-peaks with `NaN`, but then you'd need to use `nanmean` or `nanmax` instead of `mean` or `max`; this doesn't generalize to arbitrary aggregators.

Some questions for you:

* Is the slowdown common to all aggregate functions, or just weighted average?
* Have you benchmarked it against a non-masked implementation?

One other thing that I just noticed yesterday is that the output of `salience` is a masked array, not an ndarray.  Can you patch this to convert back to `ndarray`, filling with a user-supplied value for non-peak entries?  Or maybe a sparse matrix would be more appropriate?

**Comment by bmcfee** [comment]
Is this fixed by merging #458 ?

**Comment by bmcfee** [comment]
I just benched `salience` against `harmonics`, and I think we're done here.

At 22050 Hz:
```python
In [15]: S.shape
Out[15]: (1025, 2647)

In [16]: %timeit librosa.harmonics(S, librosa.fft_frequencies(),
   ....:                           h_range=[1./3, 1./2, 1, 2, 3])
1 loop, best of 3: 301 ms per loop

In [17]: %timeit librosa.salience(S, librosa.fft_frequencies(),
   ....:                          h_range=[1./3, 1./2, 1, 2, 3], fill_value=0)
1 loop, best of 3: 468 ms per loop
```

At 44100 Hz:
```python
In [20]: S.shape
Out[20]: (1025, 5294)

In [21]: %timeit librosa.harmonics(S, librosa.fft_frequencies(sr=sr),
   ....:                           h_range=[1./3, 1./2, 1, 2, 3])
1 loop, best of 3: 392 ms per loop

In [22]: %timeit librosa.salience(S, librosa.fft_frequencies(sr=sr),
   ....:                          h_range=[1./3, 1./2, 1, 2, 3], fill_value=0)
1 loop, best of 3: 652 ms per loop
```

and with different aggregators:
```python
In [23]: %timeit librosa.salience(S, librosa.fft_frequencies(sr=sr),
   ....:                           h_range=[1./3, 1./2, 1, 2, 3], fill_value=0,
   ....:                           aggregate=np.mean)
1 loop, best of 3: 562 ms per loop

In [24]: %timeit librosa.salience(S, librosa.fft_frequencies(sr=sr),
   ....:                           h_range=[1./3, 1./2, 1, 2, 3], fill_value=0,
   ....:                           aggregate=np.average)
1 loop, best of 3: 661 ms per loop

In [25]: %timeit librosa.salience(S, librosa.fft_frequencies(sr=sr),
   ....:                           h_range=[1./3, 1./2, 1, 2, 3], fill_value=0,
   ....:                           aggregate=np.max)
1 loop, best of 3: 569 ms per loop
```
so the added overhead of harmonic summation is about half that of computing harmonics in the first place.  This seems reasonable to me.

@rabitt unless you object, I'll close this one out.
